### PR TITLE
[release/7.0] Fix endless loop in illink analyzer

### PR DESCRIPTION
**Pull request by vitek-karas**
This is a port of the 8.0 change: https://github.com/dotnet/runtime/pull/86449

I verified that the repro from https://github.com/dotnet/runtime/issues/86379 is fixed with this change.

This is a 7.0 fix of https://github.com/dotnet/runtime/issues/86379

## Impact

If the right pattern occurs in the analyzed code, the analyzer will hang. This leads to hanging the build as well as causing problems with IDEs relying on a language service with the analyzer.

## Workaround

The only workaround is to disable the trim analyzer on the affected project - which means losing all of the diagnostics it provides.

## Risk

The fix is slightly risky in that it changes how some internal state of the analyzer is maintained (makes more things immutable). The fix has been merged into .NET 8 main for about 2 weeks now and we haven't seen any problems with it.

/cc @agocke

### CODE REVIEW

File: test/Mono.Linker.Tests/Tests/GetDisplayNameTests.cs [hunk 1]
```diff
@@ -191,18 +191,18 @@ public class NestedGenericClassMultipleParameters<U, V>
 		}
 
 		[DisplayName ("Mono.Linker.Tests.GetDisplayNameTests.MethodWithGenericTypeArgument(IList<GetDisplayNameTests.GenericClassOneParameter<Byte*[]>>)")]
-		public static void MethodWithGenericTypeArgument (IList<GenericClassOneParameter<byte*[]>> p)
+		public unsafe static void MethodWithGenericTypeArgument (IList<GenericClassOneParameter<byte*[]>> p)
```
Comment (sbomer) [inline test/Mono.Linker.Tests/Tests/GetDisplayNameTests.cs:194]: Just curious - what's the reason for this change?

**Review by sbomer — APPROVED** [review]

File: test/Mono.Linker.Tests/Tests/GetDisplayNameTests.cs (hunk 1 above)
Comment (vitek-karas) [reply test/Mono.Linker.Tests/Tests/GetDisplayNameTests.cs:194]: Compiler reported a warning on this - I assume it's me using a new version of the compiler.
